File: backend/app/api/v1/endpoints/exercises.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.v1.dependencies import get_current_user
from app.core import ExerciseNotFoundError
from app.core.database import get_db
from app.crud.exercise import get_exercise, list_exercises
from app.models import BodyRegion, Exercise, ExperienceLevel, MovementPattern, User
from app.schemas import ExerciseResponse

router = APIRouter(prefix="/exercises", tags=["exercises"])


@router.get("", response_model=list[ExerciseResponse])
async def list_exercises_endpoint(
    body_region: BodyRegion | None = None,
    movement_pattern: MovementPattern | None = None,
    muscle_group: str | None = None,
    equipment_tags: list[str] = Query([]),
    difficulty_level: ExperienceLevel | None = None,
    exclude_contraindications: list[str] = Query([]),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> list[Exercise]:
    exercises = await list_exercises(
        db,
        body_region=body_region,
        movement_pattern=movement_pattern,
        muscle_group=muscle_group,
        equipment_tags=equipment_tags or None,
        difficulty_level=difficulty_level,
        exclude_contraindications=exclude_contraindications or None,
    )
    return exercises


@router.get("/{exercise_id}", response_model=ExerciseResponse)
async def get_exercise_endpoint(
    exercise_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> Exercise:
    exercise = await get_exercise(db, exercise_id)
    if exercise is None:
        raise ExerciseNotFoundError()
    return exercise

File: frontend/src/api/exercises.ts
import { apiClient } from '@/api/client';
import type { ExerciseResponse } from '@/types/exercise';

export async function listExercises(params?: {
  bodyRegion?: string;
  movementPattern?: string;
  muscleGroup?: string;
  equipmentTags?: string[];
  difficultyLevel?: string;
}): Promise<ExerciseResponse[]> {
  const queryParams = new URLSearchParams();

  if (params?.bodyRegion) queryParams.append('body_region', params.bodyRegion);
  if (params?.movementPattern) queryParams.append('movement_pattern', params.movementPattern);
  if (params?.muscleGroup) queryParams.append('muscle_group', params.muscleGroup);
  if (params?.difficultyLevel) queryParams.append('difficulty_level', params.difficultyLevel);

  if (params?.equipmentTags && params.equipmentTags.length > 0) {
    params.equipmentTags.forEach((tag) => queryParams.append('equipment_tags', tag));
  }

  const queryString = queryParams.toString();
  const url = queryString ? `/exercises?${queryString}` : '/exercises';

  const { data } = await apiClient.get<ExerciseResponse[]>(url);
  return data;
}

export async function getExercise(id: number): Promise<ExerciseResponse> {
  const { data } = await apiClient.get<ExerciseResponse>(`/exercises/${id}`);
  return data;
}
